### ISSUE TITLE
[Bug] Unable to save empty input with relation HasOne

### ISSUE THREAD

**Issue by CinexUA** [Laravel-Backpack/CRUD#3944]
# Bug report

### What I did
Sometimes it is necessary to delete data from input with relation `HasOne`

I try to store at least one input blank from the list: 
```
$this->crud->addFields([
            [
                'name' => 'userInfo.first_name',
                'label' => __('First name'),
                'type' => 'text',
                'tab' => __('User info'),
                'wrapper' => [
                    'class' => 'form-group col-xs-12 col-sm-12 col-md-6 col-lg-6 col-xl-3',
                ],
            ],[
                'name' => 'userInfo.last_name',
                'label' => __('Last name'),
                'type' => 'text',
                'tab' => __('User info'),
                'wrapper' => [
                    'class' => 'form-group col-xs-12 col-sm-12 col-md-6 col-lg-6 col-xl-3',
                ],
            ],[
                'name' => 'userInfo.hrp_number',
                'label' => __('HRP - number'),
                'type' => 'text',
                'tab' => __('User info'),
                'wrapper' => [
                    'class' => 'form-group col-xs-12 col-sm-12 col-md-6 col-lg-6 col-xl-3',
                ],
            ],[
                'name' => 'userInfo.approved_hrp',
                'label' => __('HRP approved'),
                'type' => 'select_from_array',
                'tab' => __('User info'),
                'options' => UserInfo::getStatusValues(),
                'allows_null' => false,
                'default' => UserInfo::STATUS_APPROVE_PENDING,
                'wrapper' => [
                    'class' => 'form-group col-xs-12 col-sm-12 col-md-6 col-lg-6 col-xl-3',
                ],
            ],[
                'name' => 'userInfo.address',
                'label' => __('Address'),
                'type' => 'text',
                'tab' => __('User info'),
                'wrapper' => [
                    'class' => 'form-group col-xs-12 col-sm-12 col-md-6 col-lg-6',
                ],
            ],[
                'name' => 'userInfo.private_mobile',
                'label' => __('Private mobile'),
                'type' => 'text',
                'tab' => __('User info'),
                'wrapper' => [
                    'class' => 'form-group col-xs-12 col-sm-12 col-md-6 col-lg-6 col-xl-3',
                ],
            ],[
                'name' => 'userInfo.date_of_birth',
                'label' => __('Birthday'),
                'type' => 'date',
                'format' => 'd/m/Y',
                'tab' => __('User info'),
                'wrapper' => [
                    'class' => 'form-group col-xs-12 col-sm-12 col-md-6 col-lg-6 col-xl-3',
                ],
            ],
        ]);
```

### What I expected to happen

I expected to see an empty field when I delete data from it

### What happened

Data in input can only be written, changed but not deleted. I want to be able to delete data from the input field.

### What I've already tried to fix it

I was looking for a solution on the Internet, in the official documentation, stackowerflow and other resources, but could not find anything

### Is it a bug in the latest version of Backpack?

After I run ```composer update backpack/crud``` the bug... is it still there?
Yes

### Backpack, Laravel, PHP, DB version

When I run ```php artisan backpack:version``` the output is:

### PHP VERSION:
PHP 8.0.10 (cli) (built: Sep  3 2021 17:04:26) ( NTS )
Copyright (c) The PHP Group
Zend Engine v4.0.10, Copyright (c) Zend Technologies

### LARAVEL VERSION:
v8.70.2@dec9524cd0f9fa35a6eb8e25d0b40f8bbc8ec225

### BACKPACK VERSION:
4.1.60@bec5cfa11f0cd0712c79256748acf5850080c0ee

**Comment by pxpm** [comment]
Hey @tabacitu 

I was totally aware of this. And I already tried to fix it, but it would be a major BC. 

Problem is here: https://github.com/Laravel-Backpack/CRUD/blob/e034799c7493b1920ae5f946a81d6a26243f7351/src/app/Library/CrudPanel/Traits/Create.php#L221

The check for `is_null` removes the field `hasOne.some_field` from the relation data if empty because the data already went through `ConvertEmptyStringsToNull` .

On submission:
![image](https://user-images.githubusercontent.com/7188159/142607837-312e8c29-449e-418c-9ba3-efdbd4ae840c.png)

On controller:
![image](https://user-images.githubusercontent.com/7188159/142607973-bb064909-37cb-4b6e-b54b-1aa4fb6f9a83.png)

This is connected with the saving process for multiple entries somehow, because I broke it when trying to fix this and we needed to revert the PR.

In relationship branch I removed this `is_null` check, but then is necessary to add to all select multiples either the switch between `single` and `multiple` to be posted **OR** since we have repeatable as array now, I think we can go back to the old old trick of having an hidden input that gets sent when the the field is cleared. We would then need to remove the fields that are not in the request from saving so people can use `disabled` fields, and they would not be updated when not present in the request. 

The solution I've found goes along this lines:

```php
private function getRelationDataFromFormData($data)
    {
        $relation_fields = $this->getRelationFields();

        $relation_fields = $this->parseRelationFieldNamesFromHtml($relation_fields);

        //remove fields that are not in the submitted form.
        $relation_fields = array_filter($relation_fields, function ($item) use ($data) {
            return Arr::has($data, $item['name']);
        });

        $relationData = [];
        foreach ($relation_fields as $relation_field) {
            $attributeKey = $relation_field['name'];
            if (isset($relation_field['pivot']) && $relation_field['pivot'] !== true) {
```

Let me know what you think. I think we cannot fix it without BC because we cannot run the inverse of `ConvertEmptyStrings` to make it "empty string" again because we would not know what fields in the data should be converted and what should not. 

Removing the `is_null` leads to another BC, and the fix is another BC. 

Pedro

**Comment by tabacitu** [comment]
Aahh 😡  It's that damn `ConvertsEmptyStringsToNull` middleware again... 😅  

![](https://media4.giphy.com/media/eSMZFmPb6pbk4/giphy.gif)

Ok so... here's a wild idea @pxpm ... what if we... _skip that middleware when we want to_ 👀   ? Previously that was impossible (or at least very difficult), but starting with Laravel 8.36 (april 2021), the middleware has a `skipWhen()` method. I've put together PR https://github.com/Laravel-Backpack/CRUD/pull/3950 - let me know what you think about it there @pxpm 

This is either a brilliant idea, or a very stupid one. Can't figure out which one it is at the moment... 😅

**Comment by kiddtang** [comment]
@CinexUA another idea that I workaround for this issue. 
As I know this field can be cleared, I reverse `ConvertsEmptyStringsToNull` with the method

```
protected function handleEmptyInput($request)
{
    if (is_null($request->input('profile.address_2'))) {
        $profile = $request->input('profile');
        $profile['address_2'] = '';
        $request->request->set('profile', $profile);
    }
    return $request;
}
```

Inspired by `handlePasswordInput`, in [`UserCrudController.php`](https://github.com/Laravel-Backpack/PermissionManager/blob/master/src/app/Http/Controllers/UserCrudController.php)
I tried `$request->request->set('profile.address_2', '');` but it is not working. So, I just working on the array.

Apply it like `handlePasswordInput`
```
public function update()
{
    $this->crud->setRequest($this->crud->validateRequest());
    $this->crud->setRequest($this->handlePasswordInput($this->crud->getRequest()));
    $this->crud->setRequest($this->handleEmptyInput($this->crud->getRequest()));
    $this->crud->unsetValidation(); // validation has already been run

    return $this->traitUpdate();
}
```

Hope it helps :-)

**Comment by tabacitu** [comment]
Heads-up - this has been fixed in https://github.com/Laravel-Backpack/CRUD/pull/4066

Starting with Backpack 4.2, you can use HasOne and MorphOne relationships in two ways:

```php
// add text/number/etc fields for each related attribute
CRUD::field('userInfo.first_name');
CRUD::field('userInfo.last_name');

// add one field for all attributes (it will load a repeatable field with one item behind the scenes)
CRUD::field('userInfo')->subfields([
    ['name' => 'first_name', ['type' => 'text']],
    ['name' => 'last_name', ['type' => 'text']],
]);
```

The first method will NOT allow the admin to ever delete the related entry. Just update it.
But the second method will. Just click "X" on that entry and Backpack will be delete it from the db.

🎉 So I'll be closing this. If after the 4.2 upgrade you're still having this issue, please reply here again, we'll re-open and re-investigate.

**Comment by CinexUA** [comment]
Hi! When will laravel backpack 4.2 update be released? I don't want to upgrade to version 5.0 just to avoid this bug.

**Comment by tabacitu** [comment]
Hi @CinexUA ,

In short, **Backpack v4.2 == Backpack 5.0**. But it should be a free upgrade for you, if you purchased Backpack after Feb 9th 2021. You can [read more about that here](https://backpackforlaravel.com/docs/5.x/release-notes). And the upgrade process itself shouldn't be a long one at all.

Cheers!
